### PR TITLE
Pascal/sagemaker

### PR DESCRIPTION
**Pull request by nickfogler**

### CODE REVIEW

**Comment by nickfogler** [comment]
@pascal4mile, @camirmas is going to give this code a quick once-over.  I created this PR to allow any internal discussion prior to pushing to Looker

File: src/actions/sagemaker/sagemaker_infer.ts [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+      label: "SMTP Pass",
+      required: true,
+      sensitive: false,
+      description: "Pass for sending emails.",
+    },
+  ]
+
+  async execute(request: Hub.ActionRequest) {
+
+    logJson("request", request)
```
Comment (camirmas) [inline src/actions/sagemaker/sagemaker_infer.ts]: I haven't noticed many debug statements in the core repo, should these be removed before a PR to master?

File: src/actions/sagemaker/sagemaker_infer.ts [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+      // create transform job
+      const client = this.getSageMakerClientFromRequest(request)
+
+      const s3InputPath = `s3://${bucket}/${inputDataKey}`
+      const s3OutputPath = `s3://${bucket}`
+      winston.debug("s3InputPath", s3InputPath)
+      winston.debug("s3OutputPath", s3OutputPath)
+
+      const transformParams: SageMaker.CreateTransformJobRequest = {
```
Comment (camirmas) [inline src/actions/sagemaker/sagemaker_infer.ts:151]: I suppose if you wanted to cut the file size/function sizes down a bit, and DRY up some code between this and the other sagemaker actions, you could factor out the sagemaker and s3 configuration/client code

File: src/actions/sagemaker/sagemaker_infer.ts [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        required: true,
+        options: buckets.map((bucket) => {
+          return {
+            name: bucket.Name!,
+            label: bucket.Name!,
+          }
+        }),
+        type: "select",
+        // default: buckets[0].Name, // DNR
+        default: "looker-marketing-analysis", // DNR
```
Comment (camirmas) [inline src/actions/sagemaker/sagemaker_infer.ts:237]: Any particular reason to hardcode this one? Forgive me for not being familiar with this

File: src/actions/sagemaker/sagemaker_infer.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        label: "Number of instances",
+        name: "numInstances",
+        default: "1",
+        description: "The number of instances to run. Valid values: 1 to 500.",
+      },
+    ]
+    return form
+  }
+
+  protected getSageMakerClientFromRequest(request: Hub.ActionRequest) {
```
Comment (camirmas) [inline src/actions/sagemaker/sagemaker_infer.ts:265]: I'm still familiarizing with TypeScript, is it typical to allow type inference for return types when we're sure we won't get a different result (i.e. `undefined`)?

File: src/actions/sagemaker/algorithm_hosts.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  "ap-south-1": "991648021394.dkr.ecr.ap-south-1.amazonaws.com",
+  "ap-southeast-1": "475088953585.dkr.ecr.ap-southeast-1.amazonaws.com",
+  "ap-southeast-2": "712309505854.dkr.ecr.ap-southeast-2.amazonaws.com",
+  "ca-central-1": "469771592824.dkr.ecr.ca-central-1.amazonaws.com",
+  "eu-central-1": "664544806723.dkr.ecr.eu-central-1.amazonaws.com",
+  "eu-west-1": "438346466558.dkr.ecr.eu-west-1.amazonaws.com",
+  "eu-west-2": "644912444149.dkr.ecr.eu-west-2.amazonaws.com",
+}
+
+export const ecrHosts: { [key: string]: { [key: string]: string } } = {
```
Comment (camirmas) [inline src/actions/sagemaker/algorithm_hosts.ts:38]: Will these always be the same, or will they at some point need to be changed? If they need to be changed, is there a way to make them configurable?

File: src/actions/sagemaker/test_sagemaker.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// import * as chai from "chai"
```
Comment (camirmas) [inline src/actions/sagemaker/test_sagemaker.ts:1]: Is testing required for MVP?

**Review by camirmas — COMMENTED** [review]
I don't have enough context to approve/reject here feature-wise, but for the most part it looks good to me. This integration also appears to be significantly more complicated than the others, so it's difficult to provide insights on implementation and testing.

**Comment by pascal4mile** [comment]
@nickfogler I'm addressing these issues now. But we should re-target it to the looker repo (or close this PR and open a new one), and then update the master branch of our fork once they merge the PR out to their master branch.

File: src/actions/sagemaker/algorithm_hosts.ts (hunk 5 above)
Comment (pascal4mile) [reply src/actions/sagemaker/algorithm_hosts.ts:38]: adding a comment with an AWS docs URL where these lists live.

File: src/actions/sagemaker/aws_instance_types.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "ml.c4.2xlarge",
+    "ml.c4.4xlarge",
+    "ml.c4.8xlarge",
+    "ml.p3.2xlarge",
+    "ml.p3.8xlarge",
+    "ml.p3.16xlarge",
+    "ml.p2.xlarge",
+    "ml.p2.8xlarge",
+    "ml.p2.16xlarge",
+]
```
Comment (pascal4mile) [inline src/actions/sagemaker/aws_instance_types.ts:32]: adding a comment with an AWS docs URL where this list was derived from

File: src/actions/sagemaker/mail_transporter.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+//             subject: `test message ${new Date()}`,
+//             text: "This is a test message.",
+//         })
+//         console.log(result)
+//     } catch (err) {
+//         console.error(err)
+//     }
+// }
+
+// main()
```
Comment (pascal4mile) [inline src/actions/sagemaker/mail_transporter.ts]: removing this commented code

File: src/actions/sagemaker/sagemaker_infer.ts (hunk 1 above)
Comment (pascal4mile) [reply src/actions/sagemaker/sagemaker_infer.ts]: paring out most of these debug statements

File: src/actions/sagemaker/sagemaker_infer.ts (hunk 3 above)
Comment (pascal4mile) [reply src/actions/sagemaker/sagemaker_infer.ts:237]: my mistake. these are marked DNR = do not release. removing.

File: src/actions/sagemaker/sagemaker_infer.ts (hunk 4 above)
Comment (pascal4mile) [reply src/actions/sagemaker/sagemaker_infer.ts:265]: yeah i'm still learning it myself, but I think TS is able to tell that this function always returns a SageMaker instance so no need to declare that in the function signature.

File: src/actions/sagemaker/test_sagemaker.ts (hunk 6 above)
Comment (pascal4mile) [reply src/actions/sagemaker/test_sagemaker.ts:1]: yeah probably. need to get some input from wil or someone at looker about what testing is required here

**Comment by pascal4mile** [comment]
@camirmas thanks for the review! going to close this PR and open a new one to the upstream repo.

**Pull request closed without merging**